crates/iota-ledger/src/api/helpers.rs

```diff
@@ -0,0 +1,231 @@
+        vec![self as u8]
+    }
+}
+
+#[derive(Debug)]
+struct BlockResponse {
+    instruction: LedgerToHost,
+    payload: Vec<u8>,
+}
+
+impl BlockResponse {
+    fn chunk_hash(&self) -> Result<Digest<32>, errors::LedgerError> {
+        match self.instruction {
+            LedgerToHost::GetChunk => {
+                if self.payload.len() >= 32 {
+                    let mut hash = [0u8; 32];
+                    hash.copy_from_slice(&self.payload[..32]);
+                    return Ok(Digest::<32>::new(hash));
+                }
+                Err(errors::LedgerError::BlocksProtocolFailed)
+            }
+            LedgerToHost::PutChunk => Ok(Sha256::digest(&self.payload)),
+            _ => Err(errors::LedgerError::BlocksProtocolFailed),
+        }
+    }
+}
+
+impl Unpackable for BlockResponse {
+    fn unpack<R: Read>(buf: &mut R) -> Result<Self, PackableError> {
+        let instruction =
+            LedgerToHost::try_from(u8::unpack(buf)?).map_err(|_| PackableError::InvalidVariant)?;
+
+        let mut payload = Vec::new();
+        buf.read_to_end(&mut payload)?;
+        Ok(Self {
+            instruction,
+            payload,
+        })
+    }
+}
+
+pub(crate) fn send_with_blocks<T: Transport, R: Unpackable>(
+    transport: &T,
+    ins: constants::APDUInstructions,
+    payloads: Vec<Box<dyn PackableObject>>,
+    extra_data: Option<HashMap<Digest<32>, Vec<u8>>>,
+) -> Result<R, errors::LedgerError> {
+    const CHUNK_SIZE: usize = 180;
+
+    let mut data = extra_data.unwrap_or_default();
+    let mut parameter_list: Vec<Digest<32>> = Vec::new();
+
+    for payload in payloads {
+        let packed = payload
+            .pack_as_vec()
+            .map_err(|_| errors::LedgerError::Serialization)?;
+        let chunks: Vec<&[u8]> = packed.chunks(CHUNK_SIZE).collect();
+
+        let mut last_hash: Digest<32> = Digest::<32>::new([0u8; 32]);
+        for chunk in chunks.iter().rev() {
+            let mut linked_chunk = Vec::with_capacity(32 + chunk.len());
+            linked_chunk.extend(last_hash.to_vec());
+            linked_chunk.extend_from_slice(chunk);
+
+            last_hash = Sha256::digest(&linked_chunk);
+            data.insert(last_hash, linked_chunk);
+        }
+
+        parameter_list.push(last_hash);
+    }
+
+    let mut initial_payload = vec![HostToLedger::Start as u8];
+    for param in parameter_list {
+        initial_payload.extend(&param.to_vec());
+    }
+
+    handle_blocks_protocol(transport, ins, initial_payload, data)
+}
+
+fn handle_blocks_protocol<T: Transport, U: Unpackable>(
+    transport: &T,
+    ins: constants::APDUInstructions,
+    mut payload: Vec<u8>,
+    mut data: HashMap<Digest<32>, Vec<u8>>,
+) -> Result<U, errors::LedgerError> {
+    let mut result = Vec::new();
+    let ins = ins as u8;
+
+    loop {
+        let cmd = APDUCommand {
+            cla: constants::APDU_CLA,
+            ins,
+            p1: constants::APDU_P1,
+            p2: constants::APDU_P2,
```

Comment:
You are sometimes using these constants and sometimes hardcoded 0, is there a reason or some some inconsistency maybe arising from ledger.rs?

Reply:
No reason, there weren't constants in the bolos module.
Done in https://github.com/iotaledger/iota/pull/7620/commits/5ece832f61bd1369d30e9ed7fd17148971f086ef